API/Auth/DlcsBasicAuthenticationHandler.cs

```diff
@@ -0,0 +1,147 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Security.Claims;
+using System.Text;
+using System.Text.Encodings.Web;
+using System.Threading.Tasks;
+using API.Client;
+using DLCS.Model.Customers;
+using DLCS.Web.Auth;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace API.Auth
+{
+    /// <summary>
+    /// Auth in DLCS API
+    ///  - Some paths, such as https://api.dlcs.io/ and various shared policies, can serve GET requests
+    ///    to an anonymous user. Another example: https://api.dlcs.io/originStrategies
+    ///  - An admin API key is allowed to make any valid API call.
+    ///  - A customer path is one that starts /customers/n/, where n is the customer id.
+    ///    Customer-specific resources such as images, auth services, etc live under these paths.
+    ///  - Only an admin key or a key belonging to that customer can call these paths.
+    ///  - A customer path is never anonymous.
+    /// 
+    /// Care should be taken not to confuse the customer whose resource is being called with the
+    /// customer making the call.
+    /// When the caller is admin, these can be different.
+    /// The resource might not be associated with any customer.
+    /// </summary>
+    public class DlcsBasicAuthenticationHandler : AuthenticationHandler<BasicAuthenticationOptions>
+    {
+        private readonly ICustomerRepository customerRepository;
+        private readonly DeliveratorApiAuth deliveratorApiAuth; // TODO - will become apiAuth
+
+        public DlcsBasicAuthenticationHandler(
+            IOptionsMonitor<BasicAuthenticationOptions> options,
+            ILoggerFactory logger,
+            UrlEncoder encoder,
+            ISystemClock clock,
+            ICustomerRepository customerRepository,
+            DeliveratorApiAuth deliveratorApiAuth)
+            : base(options, logger, encoder, clock)
+        {
+            this.customerRepository = customerRepository;
+            this.deliveratorApiAuth = deliveratorApiAuth;
+        }
+
+        protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
+        {
+            // skip authentication if endpoint has [AllowAnonymous] attribute
+            // Not all API paths require auth...
+            var endpoint = Context.GetEndpoint();
+            if (endpoint?.Metadata?.GetMetadata<IAllowAnonymous>() != null)

```

Comment:
Is this necessary? I thought that `[AllowAnonymous]` would mean that this scheme wasn't called at all?